### PR TITLE
`null` input value support (as literal and variable value)

### PR DESCRIPTION
**Pull request by bruce**
This PR adds support for the `null` input value (as a literal and variable value) [as laid out in the October 2016 GraphQL Specification](http://facebook.github.io/graphql/#sec-Null-Value).

Test:
- [x] Comprehensive test case (now putting new specification related tests in `test/graphql_specification`, future refactoring will pull tests out of `absinthe_test.exs`, etc)

Document preparation:
- [x] Parser support (lexer, parser, AST node)
- [x] Blueprint Node, conversion

Data handling:
- [x] Variable value parsing (`nil` -> `Blueprint.Input.Null.t`)
- [x] Override nullable input object field defaults
- [x] Override nullable argument defaults
- [x] Override nullable variable defaults

Resolution:
- [x] Pass as `nil` to resolvers

### CODE REVIEW

File: lib/absinthe/phase/document/variables.ex [hunk 1]
```diff
@@ -76,4 +72,19 @@ defmodule Absinthe.Phase.Document.Variables do
     {node, acc}
   end
 
+  defp get(map, key, default) do
+    if Map.has_key?(map, key) do
+      Map.get(map, key) |> Blueprint.Input.parse
+    else
+      case {Map.get(map, key), default} do
```
Comment (jparise) [inline lib/absinthe/phase/document/variables.ex]: Could this be simplified by passing the default value directly to `Map.get`?

```elixir
case Map.get(map, key, default) do
  nil -> nil
  value -> Blueprint.Input.parse(value)
end
```

I think that covers all of the same cases.

Comment (jparise) [inline lib/absinthe/phase/document/variables.ex]: Ah, I see now. It's that last case (`nil` stored as an explicit Map value) that I wasn't considering. Thanks for the explanation!

**Comment by bruce** [comment]
- All tests (675) now pass in this PR
- An additional validation, `ProvidedNonNullVariables` has been added

Note the following [arguably incompatible] change: When `null` is provided as a literal or a variable value, `nil` values do make it all the way to resolvers (note that an explicit `null`/`nil` value supersedes argument default values). We should probably think about whether this warrants more than a patch version bump.

Please review for merge and release. (@benwilson512, I suggest we try this against our application.)

File: lib/absinthe/phase/document/arguments/data.ex [hunk 2]
```diff
@@ -49,13 +49,20 @@ defmodule Absinthe.Phase.Document.Arguments.Data do
   def handle_node(%Input.Value{normalized: %Input.Object{fields: fields}} = node) do
     data =
       fields
-      |> Enum.map(&{&1.schema_node.__reference__.identifier, &1.input_value.data})
-      |> Enum.reject(&match?({_, nil}, &1))
+      |> Enum.filter_map(
+        &include_field?/1,
+        &{&1.schema_node.__reference__.identifier, &1.input_value.data}
+      )
```
Comment (benwilson512) [inline lib/absinthe/phase/document/arguments/data.ex]: Nice, this is cleaner than map / reject.

**Comment by bruce** [comment]
Good point on inclusion in the list. Adding some tests and working through them now.

Minimum cases we need to handle (using `Int` as the example element type):

- `[Int]`
- `[Int!]`
- `[Int]!`
- `[Int!]!`

Given the following values:

- `null`
- `1`
- `[]`
- `[1]`
- `[null, 1]`

**Comment by bruce** [comment]
@benwilson512 I think this is ready for another review pass. The above usage of `null` (both as literals and variable values) in conjunction with lists have been tested. This PR adds ~30 new tests for various permutations.

**Comment by bruce** [comment]
Switching this to target `next` and merging.

**Comment by bruce** [comment]
Note: We're planning this as part of the `v1.4` release. We're bundling some other breaking changes to go along with it in the `next` branch.

**Comment by zipme** [comment]
Hi, this doesn't seem to be working correctly with enum type.

I have this enum type defined:
```elixir
enum :transaction_type do
  value :in
  value :out
end
```
and 
```elixir
connection field :transactions, node_type: :transaction do
    arg :transaction_type, :transaction_type
    ...
end
```

when query with:

```elixir
query  {
  transactions(first: 10, transactionType: null) {
    edges {
      node {
        id
      }
    }
  }
}
```

I got this error 
```json
{
  "errors": [
    {
      "message": "Argument \"transactionType\" has invalid value null.",
      "locations": [
        {
          "line": 2,
          "column": 0
        }
      ]
    }
  ]
}
```

**Comment by benwilson512** [comment]
Hey @zipme that does indeed look like a bug, please open a dedicated issue.

**Pull request merged**